File: apk/src/com/android/healthconnect/controller/shared/preference/HealthExpandablePreference.java
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.healthconnect.controller.shared.preference;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.ImageView;

import androidx.preference.Preference;
import androidx.preference.PreferenceGroup;
import androidx.preference.PreferenceViewHolder;

import com.android.healthconnect.controller.R;

/**
 * A {@link PreferenceGroup} that can be expanded and collapsed.
 *
 * <p>This preference will display an arrow that can be clicked to expand or collapse the
 * preferences contained within it.
 */
public class HealthExpandablePreference extends PreferenceGroup {

    private boolean mIsExpanded = false;
    private OnExpandChangeListener mOnExpandChangeListener;

    /**
     * Interface definition for a callback to be invoked when the expansion state of this preference
     * changes.
     */
    public interface OnExpandChangeListener {
        /**
         * Called when the expansion state of this preference has changed.
         *
         * @param isExpanded The new expansion state.
         */
        void onExpandChanged(boolean isExpanded);
    }

    public HealthExpandablePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        setLayoutResource(R.layout.expandable_preference);
    }

    @Override
    public void onBindViewHolder(PreferenceViewHolder holder) {
        super.onBindViewHolder(holder);
        holder.setDividerAllowedBelow(false);
        holder.setDividerAllowedAbove(false);

        ImageView arrow = (ImageView) holder.findViewById(R.id.expand_arrow);
        arrow.setRotation(mIsExpanded ? 180f : 0f);
        holder.itemView.setOnClickListener(
                v -> {
                    setExpanded(!mIsExpanded);
                    if (mOnExpandChangeListener != null) {
                        mOnExpandChangeListener.onExpandChanged(mIsExpanded);
                    }
                });
        updatePreferences();
    }

    @Override
    public boolean addPreference(Preference preference) {
        final boolean result = super.addPreference(preference);
        updatePreferences();
        return result;
    }

    /** Sets the expansion state of this preference. */
    public void setExpanded(boolean isExpanded) {
        if (mIsExpanded != isExpanded) {
            mIsExpanded = isExpanded;
            updatePreferences();
            notifyChanged();
        }
    }

    /** Sets a callback to be invoked when the expansion state of this preference changes. */
    public void setOnExpandChangeListener(OnExpandChangeListener listener) {
        mOnExpandChangeListener = listener;
    }

    private void updatePreferences() {
        for (int i = 0; i < getPreferenceCount(); i++) {
            getPreference(i).setVisible(mIsExpanded);
        }
    }
}

File: apk/src/com/android/healthconnect/controller/shared/preference/HealthExpandablePreference.kt
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.healthconnect.controller.shared.preference

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.ImageView
import androidx.appcompat.widget.SwitchCompat
import androidx.core.view.isVisible
import androidx.preference.PreferenceViewHolder
import com.android.healthconnect.controller.R
import com.android.settingslib.widget.SettingsThemeHelper

class HealthExpandablePreference
@JvmOverloads
constructor(context: Context, attrs: AttributeSet? = null) :
    BaseExpandablePreference(context, attrs) {

    private var onCheckedChangeListener: ((Boolean) -> Unit)? = null
    var isChecked: Boolean = false
        set(value) {
            if (field != value) {
                field = value
                notifyChanged()
            }
        }

    var switchContentDescription: CharSequence? = null
        set(value) {
            field = value
            notifyChanged()
        }

    init {
        layoutResource =
            if (SettingsThemeHelper.isExpressiveTheme(context)) {
                R.layout.preference_expressive_expandable_switch
            } else {
                R.layout.preference_non_expressive_expandable_switch
            }
        widgetLayoutResource =
            if (SettingsThemeHelper.isExpressiveTheme(context)) {
                R.layout.expressive_expand_arrow_switch_widget
            } else {
                R.layout.non_expressive_expand_arrow_switch_widget
            }
    }

    fun setOnSwitchChangeListener(listener: ((Boolean) -> Unit)?) {
        onCheckedChangeListener = listener
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)

        val expandArrow = holder.findViewById(R.id.expand_arrow) as? ImageView
        val twoTargetDivider = holder.findViewById(R.id.two_target_divider) as? View
        val switch = holder.findViewById(R.id.switch_widget) as? SwitchCompat
        switch?.isClickable = true
        switch?.contentDescription = switchContentDescription

        switch?.setOnCheckedChangeListener(null)
        switch?.isChecked = isChecked
        switch?.setOnCheckedChangeListener { _, newIsChecked ->
            if (isChecked != newIsChecked) {
                isChecked = newIsChecked
                onCheckedChangeListener?.invoke(newIsChecked)
            }
        }

        val hasChildren = preferenceCount > 0
        expandArrow?.isVisible = hasChildren
        twoTargetDivider?.isVisible = hasChildren

        expandArrow?.rotation = if (mIsExpanded) 180f else 0f

        if (hasChildren) {
            holder.itemView.setOnClickListener {
                logger.logInteraction(
                    logName,
                    com.android.healthconnect.controller.utils.logging.UIAction.ACTION_CLICK,
                )
                setExpanded(!mIsExpanded)
                mOnExpandChangeListener?.onExpandChanged(mIsExpanded)
                notifyChanged()
            }
        } else {
            holder.itemView.setOnClickListener(null)
        }
    }

    override fun getDropDownIconId(): Int {
        return R.drawable.ic_expand_more
    }
}
